Upgrade services

rahushen opened the pull request:
Added Enhancements to existing APIs.
- Handle Backend Processing Error
- Upgrade? waits for config ready
- Rename Service to Upgrade

mikewiebe committed:
Initial upgrade services

mikewiebe committed:
Remove white space

rahushen committed:
pulling latest develop content

rahushen committed:
Merge remote-tracking branch 'origin/develop' into upgrade_services

rahushen committed:
Allows tokens to be passed to get_command

rahushen committed:
Add tests and procs for install all nxos

rahushen committed:
Update DOC to highlight that get_command can be tokenized

rahushen committed:
Add YAML to specify params for upgrade

rahushen committed:
Adding error handling and cleaning up code

rahushen committed:
Reworked some TCs + Add new TCs

rahushen committed:
Handle case when upgrade is run with booted image

rahushen committed:
Add test_upgrade_boot_image

rahushen committed:
Add Service to CHANGELOG.md

rahushen committed:
Add TODO to upgrade scenario

rahushen committed:
Update CHANGELOG.md

rahushen committed:
Merge remote-tracking branch 'origin/develop' into upgrade_services

rahushen committed:
Add command for upgrade_force

rahushen committed:
Update upgrade to handle delete_boot and force upgrade

rahushen committed:
Merge remote-tracking branch 'origin/develop' into upgrade_services

rahushen committed:
Fix test_image_version

rahushen committed:
Add is_box_online

rahushen committed:
Enhancements to service.rb

Function box_online? added.
Modified upgraded? to check for config ready.
Add version check to 'upgrade'.
Handle 500 - Backend processing Error

rahushen committed:
Enchancements added to test_service.rb

Added test for box_online?
Modified calls to Service.upgrade to include version
Increases retry to 5 in test_upgrade
Made test_upgrade_boot_image skippable in absence of yaml file

rahushen committed:
Renaming files from service to upgrade

rahushen committed:
Updating references of service to upgrade

rahushen committed:
Renaming Service to Upgrade

rahushen committed:
Updating CHANGELOG.md

mikewiebe commented on lib/cisco_node_utils/upgrade.rb:
```diff
@@ -0,0 +1,116 @@
+    # Attempts to upgrade the device to 'image'
+    def self.upgrade(version, image, media='bootflash:', del_boot=false,
+                     force_all=false)
+      # IMPORTANT - Check if version of image equals the version provided.
```
Couple things here.
1. We should change media to be uri for consistency.
2. Given that we only support bootflash at the moment, we should probably fail if anything other then bootflash is used as a uri

mikewiebe commented on tests/test_upgrade.rb:
```diff
@@ -13,88 +13,94 @@
   end
 
   def test_delete
     shell_command('touch /bootflash/foobar')
```
Just realized that `shell_command` will create a file on bootflash if we are running this in a `native` or `guestshell` environment but will not work for oac.  The `shell_command` api is just a wrapper around  `run bash` or `run guestshell` but there is no equivalent `run oac` command today.  Not sure if this will work, but see if you can replace `shell_command('touch /bootflash/foobar')` with `config("show version > bootflash:foo.txt")` instead.  You may need to escape the `>`.

rahushen committed:
Incorporated @mikewiebe PR comments.

- updated references of 'media' to 'uri'
- fail upgrade if uri is not bootflash
- replace shell_command with config to handle OAC

rahushen committed:
Fix Rubocop error

rahushen committed:
Merge remote-tracking branch 'origin/develop' into upgrade_services

mikewiebe commented:
👍
